main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -21,51 +21,670 @@ import ca.uwaterloo.flix.language.ast.{Ast, TokenErrorKind, ReadAst, SourceKind,
+      case _ if keyword("use") => TokenKind.KeywordUse
+      case _ if keyword("where") => TokenKind.KeywordWhere
+      case _ if keyword("with") => TokenKind.KeywordWith
+      case _ if keyword("without") => TokenKind.KeywordWithout
+      case _ if keyword("yield") => TokenKind.KeywordYield
+      case _ if keyword("Set#") => TokenKind.SetHash
+      case _ if keyword("Array#") => TokenKind.ArrayHash
+      case _ if keyword("Map#") => TokenKind.MapHash
+      case _ if keyword("List#") => TokenKind.ListHash
+      case _ if keyword("Vector#") => TokenKind.VectorHash
+      case _ if isMathNameChar(c) => mathName()
+      case _ if isGreekNameChar(c) => greekName()
+      // User defined operators.
+      case _ if validUserOpTokens.contains(c) =>
+        val p = peek()
+        if (validUserOpTokens.contains(p)) {
+          userDefinedOp()
+        } else if (c == '-' && p.isDigit) {
+          number() // negative numbers.
+        } else {
+          validUserOpTokens.apply(c)
+        }
+      case c if c.isLetter => name(c.isUpper)
+      case c if c.isDigit => number()
+      case '$' => builtIn()
+      case '\"' => string()
+      case '\'' => char()
+      case '`' => infixFunction()
+      case _ => TokenKind.Err(TokenErrorKind.UnexpectedChar)
+    }
+
+    addToken(kind)
+  }
+
+  /**
+   * Checks whether the following substring matches a keyword. Note that __comparison includes current__.
+   */
+  private def keyword(k: String)(implicit s: State): Boolean = {
+    // check if the keyword can appear before eof
+    if (s.current.offset + k.length > s.src.data.length) {
+      return false
+    }
+
+    val start = s.current.offset - 1
+    var matches = true
+    var offset = 0
+    while (matches && offset < k.length) {
+      if (s.src.data(start + offset) != k(offset)) {
+        matches = false
+      } else {
+        offset += 1
+      }
+    }
+    if (matches) {
+      for (_ <- 1 until k.length) {
+        advance()
+      }
+    }
+
+    matches
+  }
+
+  /**
+   * Moves current position past a built-in function, IE. "$BUILT_IN$".
+   */
+  private def builtIn()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (peek() == '$') {
+        advance()
+        return TokenKind.BuiltIn
+      }
+      advance()
+    }
+    TokenKind.Err(TokenErrorKind.UnterminatedBuiltIn)
+  }
+
+  /**
+   * Moves current position past all whitespace characters.
+   */
+  private def whitespace()(implicit s: State): Unit = {
+    while (!isAtEnd()) {
+      if (!peek().isWhitespace) {
+        return
+      }
+      advance()
+    }
+  }
+
+  /**
+   * Moves current position past a name (both upper- and lower-case).
+   * There are edge cases of variable holes, IE. "x?", and java names like "Map$Entry",
+   * which is the reason this function will return a `TokenKind`.
+   */
+  private def name(isUpper: Boolean)(implicit s: State): TokenKind = {
```

Comment:
I would call these `lexName` or `acceptName` because they return TokenKind and to distinguish them from `peek` which is a lower-level primitive.